### ISSUE TITLE
Missing UI external auth controls when CustomAuth is both enabled *and* required

### ISSUE THREAD

**Issue by chrisinmtown** [MISP/MISP#7488]
In version 2.4.143 I do not see the external auth key textbox in the admin add-user and edit-user pages when external auth is set to enabled and required via the Cake admin CLI tool.  Those pages show the set-password checkbox and password fields, which are not helpful because our users are never asked for a MISP password.

This all started when I tried to enable external authentication by editing config.php directly. I learned this was foolish, I should have used either the UI or the Cake Admin tool to change those settings. That means the first 10 or so comments in this issue just document my learning process and can mostly be skipped. 

So here goes, what did I try initially:
```
    'CustomAuth_enable' => true,
    'CustomAuth_header' => 'USER',
    'CustomAuth_name' => 'My External authentication',
    'CustomAuth_header_namespace' => 'REMOTE_',
    'CustomAuth_disable_logout' => true,
    'CustomAuth_required' => true,
```
After restarting the server, I tried to load a page after traversing my external auth method successfully, but MISP showed internal server error and wrote the following message in error.log:
```
Error: [PDOException] SQLSTATE[42S22]: Column not found: 1054 Unknown column 'User.external_auth_key' in 'where clause'
Request URL: /events/index
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(715): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1226): DboSource->fetchAll()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(3053): DboSource->read()
#5 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(3025): Model->_readDataSource()
#6 /var/www/MISP/app/Model/User.php(684): Model->find()
#7 /var/www/MISP/app/Model/User.php(674): User->getAuthUserByConditions()
#8 /var/www/MISP/app/Controller/AppController.php(1031): User->getAuthUserByExternalAuth()
#9 /var/www/MISP/app/Controller/AppController.php(1225): AppController->checkExternalAuthUser()
#10 /var/www/MISP/app/Controller/AppController.php(187): AppController->__customAuthentication()
#11 /var/www/MISP/app/Controller/EventsController.php(62): AppController->beforeFilter()
#12 /var/www/MISP/app/Lib/cakephp/lib/Cake/Event/CakeEventManager.php(244): EventsController->beforeFilter()
#13 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(682): CakeEventManager->dispatch()
#14 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(189): Controller->startupProcess()
#15 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#16 /var/www/MISP/app/webroot/index.php(92): Dispatcher->dispatch()
#17 {main}
```
I saw that AppModel.php has magic code to add columns dynamically to the database:

https://github.com/MISP/MISP/blob/81543d05d43d709a38230d208199f525cb28ab10/app/Model/AppModel.php#L363

I guess the server did not run this step in my case.

### Work environment

| Questions                 | Answers
|---------------------------|--------------------
| Type of issue             | Bug
| OS version (server)       | RedHat
| OS version (client)       | MacOS
| PHP version               | 7.4
| MISP version / git hash   | 2.4.143
| Browser                   | Firefox

### Expected behavior

Setting entries in config.php for CustomAuth turn off the MISP login screen, let me configure a user for external auth.

### Actual behavior

I only get a page Internal Server Error due to the failed database access statement.

### Steps to reproduce the behavior

Use config.default.php, add the CustomAuth lines shown above, restart the server and try to view a page; also try to add a user.

This issue MAY be a dupe of https://github.com/MISP/MISP/issues/5664 but this error didn't appear there, so I thought it might be worth asking.

I also read this carefully: https://github.com/MISP/MISP/issues/3059

Those issues mention "tick the External authentication user checkbox" but in 2.4.143 when editing a user I see no such checkbox.

**Comment by chrisinmtown** [comment]
I tried again to add a user and I caught a brief glimpse of external auth stuff in the add/edit user dialog.  I inspected the web page and found the following, the inspector shows the style="display: none;" which may explain why it's invisible!  So, ahh, why is it turned off? We have Plugin.CustomAuth_enable set to true in the system's config.php file.
```
<div class="input text"><label for="UserExternalAuthKey">
External Auth Key</label><input name="data[User][external_auth_key]" type="text" id="UserExternalAuthKey">
</div>        
```
Maybe this javascript code is not doing what's expected?
https://github.com/MISP/MISP/blob/81543d05d43d709a38230d208199f525cb28ab10/app/webroot/js/misp.js#L3746

**Comment by chrisinmtown** [comment]
I guess it's safe to say that if I want to enable custom authentication, the database must first be extended with columns to store additional information for each user. No extra columns in the db means no custom (external) auth in the server. Is there some other magic setting I must adjust to make the edit controls visible in the add/edit user dialog for the external auth data? Because I suspect that the first time I set an external auth on a user, the app will extend the users table in the database as required.

**Comment by chrisinmtown** [comment]
One more comment. In a legacy system where I *do* have custom (external) auth working, the edit-user form only shows a checkbox to set the password and accepts password strings; no controls (e.g., a checkbox and textbox) for external auth appear. I think those password fields should be disabled in this case by $password in admin_add.ctp (see below). But I could not find any use of $password in that file, so maybe something is not right.

https://github.com/MISP/MISP/blob/81543d05d43d709a38230d208199f525cb28ab10/app/View/Users/admin_add.ctp#L13

**Comment by chrisinmtown** [comment]
On gitter @iglocska suggested this might (might!) be a UI regression. The view still has the UI controls, but they are not enabled/shown as I expect. Hoping someone can confirm.

**Comment by chrisinmtown** [comment]
Turns out @github-germ reported this long before I did here: https://github.com/MISP/MISP/issues/5701

So it appears this problem has been lurking for a *long* time.

**Comment by chrisinmtown** [comment]
After some puzzling I figured out that the external auth checkbox appears if setting 'CustomAuth_enable' => true *and* 'CustomAuth_required' => false.  So the edit controls can be made visible.

But I think this is the bug: the external auth checkbox is not shown if CustomAuth_required = **true**. It sorta makes sense that you don't accept an external auth setting on a per-user basis if it's globally required. HOWEVER if that checkbox is not shown, then the text box "External Auth Key" is never shown, and without that, I cannot enter the user's auth key. 

I am trying to learn what triggers the code that dynamically adds columns `external_auth_required` and `external_auth_key` to the users table; pieces of the puzzle are here:

https://github.com/MISP/MISP/blob/81543d05d43d709a38230d208199f525cb28ab10/app/Model/Server.php#L6733

https://github.com/MISP/MISP/blob/81543d05d43d709a38230d208199f525cb28ab10/app/Model/Server.php#L2067

Is it possible that the only valid way to enable custom auth the **first** time is via the web UI -- not by editing config.php?

**Comment by chrisinmtown** [comment]
Test results - if I use the Server Settings and Maintenance page, Plugin settings, expand the CustomAuth section, then change the setting Plugin.CustomAuth_enable from false to true, after I click save the `users` table is magically extended with the two critical columns.

Next thought: Can I trigger this same action via a cake Admin setSetting invocation from the command line?

**Comment by chrisinmtown** [comment]
whew, good news, this command-line invocation does indeed trigger the database change:
```
/var/www/MISP/app/Console/cake Admin setSetting "Plugin.CustomAuth_enable" true
```

**Comment by iglocska** [comment]
> After some puzzling I figured out that the external auth checkbox appears if setting 'CustomAuth_enable' => true _and_ 'CustomAuth_required' => false. So the edit controls can be made visible.
> 
> But I think this is the bug: the external auth checkbox is not shown if CustomAuth_required = **true**. It sorta makes sense that you don't accept an external auth setting on a per-user basis if it's globally required. HOWEVER if that checkbox is not shown, then the text box "External Auth Key" is never shown, and without that, I cannot enter the user's auth key.
> 
> I am trying to learn what triggers the code that dynamically adds columns `external_auth_required` and `external_auth_key` to the users table; pieces of the puzzle are here:
> 
> https://github.com/MISP/MISP/blob/81543d05d43d709a38230d208199f525cb28ab10/app/Model/Server.php#L6733
> 
> https://github.com/MISP/MISP/blob/81543d05d43d709a38230d208199f525cb28ab10/app/Model/Server.php#L2067
> 
> Is it possible that the only valid way to enable custom auth the **first** time is via the web UI -- not by editing config.php?

Indeed, editing config.php directly will not fire the hooks, generally always edit it via the UI or via the API/CLI rather than modifying config.php directly.

**Comment by chrisinmtown** [comment]
tl;dr:   the admin add/edit user UI does not display the external auth checkbox and textbox controls if I use either the UI or Cake Admin to set settings Plugin.CustomAuth_enable = 1 **and** Plugin.CustomAuth_required = 1. I believe this is a bug. I expect those controls to appear if Plugin.CustomAuth_enable = 1, regardless of other settings.

**Comment by iglocska** [comment]
That is indeed correct though. You have arrived in that state by circumventing the built in mechanisms for enabling the features.

**Comment by chrisinmtown** [comment]
@iglocska thanks for the note. In my latest efforts I believe (!) I take the correct steps for enabling CustomAuth features.  My docker build & startup procedures use the Cake Admin command-line to set the two Plugin CustomAuth required/enable settings to 1/true (no longer starting with a config.php file with them enabled). But after setting those, the admin add-user and edit-user screens don't give me the ability to enter the user's external auth key.  If I set Plugin.CustomAuth_required to false, those edit screens change and I *can* enter a user's external auth key.

Here's an excerpt from the startup sequence of my dockerized misp:
```
misp-server_1  | Welcome to CakePHP v2.10.24 Console
misp-server_1  | ---------------------------------------------------------------
misp-server_1  | App : app
misp-server_1  | Path: /var/www/MISP/app/
misp-server_1  | ---------------------------------------------------------------
misp-server_1  | Setting "Plugin.CustomAuth_enable" changed to 1
misp-server_1  | 
misp-server_1  | 
misp-server_1  | Welcome to CakePHP v2.10.24 Console
misp-server_1  | ---------------------------------------------------------------
misp-server_1  | App : app
misp-server_1  | Path: /var/www/MISP/app/
misp-server_1  | ---------------------------------------------------------------
misp-server_1  | Setting "Plugin.CustomAuth_required" changed to 1
```
You feel that the external authkey UI controls should not appear after this? That surprises me. How is an admin supposed to enter that key when external auth is always required; i.e., users should never see the MISP login screen? Here's the top part of the admin edit-user screen showing only password even tho user passwords are never accepted when CustomAuth is required:

<img width="537" alt="Screen Shot 2021-06-15 at 8 02 55 PM" src="https://user-images.githubusercontent.com/10234212/122138433-b9fd8700-ce14-11eb-8058-845e28e61b72.png">

**Comment by iglocska** [comment]
That is indeed weird. With custom auth enabled I have this:

![image](https://user-images.githubusercontent.com/3668672/122195933-11dcd200-ce97-11eb-8299-2b9f1fc43dd9.png)

**Comment by chrisinmtown** [comment]
hi @iglocska thanks for checking on your system. Would you please try setting Plugin.CustomAuth_required = 1 and look again?  I don't know if that will break your login tho :( 

If I use Cake Admin to set Plugin.CustomAuth_required = 0 then the Admin Edit User screen looks exactly like what you show. But our local requirement is to require external auth for *all* users, so would really like to configure the system that way.

**Comment by iglocska** [comment]
Ahhh, I see indeed the logic fail in there. True, customauth required would mean that the external field is always required, however, it fails to enable it since the checkbox is no longer present (as there's no option for the user). I'll have a fix for that today, the required was the bit that I was missing.

**Comment by chrisinmtown** [comment]
👍🙏

**Comment by chrisinmtown** [comment]
A new question for you @iglocska: will you please consider disabling (or hiding) the password checkbox and fields when CustomAuth_required = 1 ?

**Comment by iglocska** [comment]
Yeah that makes sense too

**Comment by chrisinmtown** [comment]
Thank you for the quick response. I look forward to testing this!

**Comment by iglocska** [comment]
Holler if it doesn't work!

**Comment by chrisinmtown** [comment]
Thanks to docker I can build & test quickly :) I used commit adb77afaed22291cb076cf7fbb7755c3eb8ada29. Here's an image of what I see on a system with Plugin.CustomAuth_enabled = 1 and Plugin.CustomAuth_required = 1:

<img width="467" alt="Screen Shot 2021-06-16 at 1 51 59 PM" src="https://user-images.githubusercontent.com/10234212/122268512-270f2c00-ceaa-11eb-8083-746c6a5184ad.png">

I am *very glad* that the view now shows the External Auth Key field, thank you.

But I'm a little sad that the password checkbox and fields are still shown. Passwords are not used when Plugin.CustomAuth_required = 1, so I feel the password edit controls should not be shown.
